rfd/0231-in-band-mfa-ssh-sessions.md

```diff
@@ -0,0 +1,764 @@
+per-session MFA certificates.
+
+## Non-Goals
+
+1. This RFD does not propose changes to other Teleport access protocols such as Kubernetes, databases, desktops, etc.
+   However, the architecture could be extended to these protocols in the future.
+
+## Details
+
+### UX
+
+No changes are expected since this is an internal change.
+
+### High-Level Flow
+
+The client will first dial its target SSH host using the Proxy and its standard Teleport client certificate. The Proxy
+will then resolve the target cluster and host, and invoke the `EvaluateSSHAccess` RPC of the [Decision
+service](https://github.com/gravitational/Teleport.e/blob/master/rfd/0024e-access-control-decision-api.md). Upon
+receiving a permit from the Decision service, the Proxy will staple the permit to the connection and open a connection
+to the respective cluster's SSH service.
+
+The SSH service will authenticate the client using the provided client certificate and the stapled permit. The SSH
+service will then check if MFA is required for the session by examining the permit.
+
+If MFA is _not required_, the SSH service will then proceed to establish the SSH session.
+
+If MFA _is required_, the SSH service will send a Protobuf [`MFAPrompt`
+message](#ssh-keyboard-interactive-authentication) via the SSH [keyboard-interactive
+channel](https://www.rfc-editor.org/rfc/rfc4256) to inform the client that MFA is needed. The client must then call the
+`CreateChallenge` RPC on the MFA service, providing a _Session Identifying Payload (SIP)_ and any relevant metadata. The
+SIP is an SSH session hash computed from SSH session state and is used to bind the MFA challenge to the specific
+session. Both the SSH client and the SSH server can independently compute the SIP from session state.
+
+The MFA service will [store the Session Identifying Payload (SIP)](#storing-session-identifying-payloads) and respond
+with an MFA challenge for the client to solve.
+
+Next, the client solves the MFA challenge and calls `ValidateChallenge` with the MFA response.
+
+The MFA service validates the MFA response and stores the validated challenge response and SIP for temporary storage.
+
+Depending on the target cluster, the MFA service's validation step differs slightly:
+
+- **Local clusters**: Uses the `CreateValidatedChallenge` RPC internally on the same cluster.
+- **Leaf clusters**: Forwards the validated response to the leaf cluster's MFA service using the
+  `CreateValidatedChallenge` RPC.
+
+If validation is successful, the MFA service responds to the client with a confirmation that the challenge has been
+validated.
+
+The client then sends a [`MFAPromptResponse`](#ssh-keyboard-interactive-authentication) message to the SSH service,
+instructing it to retrieve the validated challenge from the MFA service.
+
+The SSH service retrieves the validated challenge from the MFA service using the `GetValidatedChallenge` RPC. The SSH
```

Comment:
I think we should make note somewhere that `GetValidatedChallenge` will function in the same manner as GetHeadlessAuthentication as suggested by @fspmarshall.

Meaning, because we need to wait for the ValidatedChallenge to exist in the backend, either our local backend, or to be replicated to the leaf backend, within the RPC handler we will create a local watcher with a timeout so that it isn't the client's responsibility to keep polling `GetValidatedChallenge` until replication is complete.